README
Added pokemon battle system. Display the current hp of your pokemon and enemy pokemon. Allows 4 different types of moves, fight, pack, run and capture.

Changelog.md
Jan 29
Added printMap(), generateMap(), generateBorder() and initializeMap()
Added generateTallGrass(), generateWater(), generateShortGrass(), generateTrees(), populateArea(), generateBoulders(),
isValidPosition(), selectPointL(), and selectPointR()
Added fillEmpty();
Added generateExit(), generateNSPath() and generateWEPath(), also added color to output
Added generateBuildings()
Moved the order of map generation so that it generates borders -> exits -> paths -> buildings -> various regions

Feb 5
Created the function to allow the keeping track of the generated maps and the moving between maps and fly function. Removed the scattering of trees and boulders during map generation due to segmentation error. Fixed some minor bugs. Also moved some of the functions from TerrainGenerator.c to utilities.h. Created a function to calculate distance of map from origin point and use the distance to determine the probability of buildings spawning.

Feb 6
Added the handling of edge maps where if it is at the edge, no exit will be generated in that direction

Feb 12
Added method getNodeCost(), getNeighbor(), dijkstra() method with reference to https://en.wikipedia.org/wiki/Dijkstra%27s_algorithm
Also added a priority queue in utilities.h from https://www.geeksforgeeks.org/priority-queue-using-linked-list/

Feb 24
Changed terrainGeneration order to generate path and buildings after generating terrains.
Changed path to be generated as bridge '=' when it is over water. 
Added another priority queue for game turn.

Feb 25
Added methods:
    generateNPC(), getNumTrainers(), setWanderer(), setExplorer(), setSwimmer(), gameTurn_init(), gameTurn_exec(), getPath();

Feb 26
Debugged the various functions for npc movement. 

Mar 12 
Added ncurses as a include in utilities.h. Added movePlayer() to allow the movement of player based on input key. Removed the use of usleep for delay. 

Mar 13
Added color to terrain. Added the function to view the npcs in the current map with the distance to them.

Mar 14
Renamed generatePlayer() to spawnPlayer(). Added the function for pokeBattle() and added the boolean isDefeated for pcStruct. npcs stop moving once they are defeated.

Mar 15 
Allow the movement of player to other maps through the exit gate.

Mar 25
Changed all files to c++
Refactored the code into multiple files to make it easier to read.

Mar 26
Added Fly function

April 1
Added classes for the different files to parse

April 2
Added the method to parse the CSV file

April 8
Added function to spawn wild pokemon in tall grass ":" with a 10% chance
Added function to display encounter with wild pokemon by displaying stats of the wild pokemon

April 9
Added function to start the game by allowing user to select their starting pokemon
Added function to assign 1-6 pokemon(s) to all trainers when generating npc
Changed pokemonBattle to call the pokemonEncounter function as a placeholder to display the stats of the first pokemon of the trainer.

April 10
Fixed a bug where Floating point error will happen when loading a new map

April 16
Implemented PokemonBattle function.

April 17
Make PokeMart and PokeCenter actually restore your item/pokemon respectively
Added Screen transition for pokemon battle
Added text animation when displaying messages to the user

util1.h
// Utilities folder
#ifndef UTIL1_H
#define UTIL1_H
#endif

#include "util5_database.h"
#include <ncurses.h>
#include <math.h>
#include <stdio.h>
#include <stdbool.h>
#include <stdlib.h>
#include <time.h>
#include <ctype.h>
#include <string.h>
#include <sys/stat.h>
#include <sys/types.h>
#include <limits.h>
#include <sys/time.h>
#include <assert.h>
#include <unistd.h>
#include <sys/ioctl.h>
#include <iostream>
#include <fstream>
#include <sstream>

using namespace std;

#define COL_COUNT 80 // number of column of terrain
#define ROW_COUNT 21 // number of row of terrain

//colors for ncurses
#define TALL_GRASS 1
#define SHORT_GRASS 2
#define FOREST 3
#define MOUNTAIN 4
#define WATER 5
#define PATH 6
#define POKEMART 7
#define POKECENTER 8
#define DEFAULT 9

//variables to hold the data parsed from csv files
std::vector<pokemon> pokemonArray;

std::vector<moves> movesArray;

std::vector<pokeMoves> pokemonMovesArray;

std::vector<pokemon_species> pokemonSpeciesArray;

std::vector<experience> expArray;

std::vector<type_names> typeNamesArray;

std::vector<pokemon_stats> pokemonStatsArray;

std::vector<stats> statsArray;
std::map<std::string, int> statsMap;

std::vector<pokemon_types> pokemonTypesArray;

/*************function stubs***************************/
void dijkstra(char terrain[ROW_COUNT][COL_COUNT], int npcType, long distanceMap[ROW_COUNT][COL_COUNT]);
 
/******************************************************/

// return the distance between 2 points
typedef struct point {
  int row;
  int col;
  char element;
} point;

typedef struct pathStruct{
    point *pathing; //array containing each point on the NPC pathing route
    int onPath; //curent spot on path
    int pathLength; //length of path
    int direction; //direction of pathing (to or from)
} pathStruct;

class itemStruct{
    public:
        int quantity;
        std::string name;
};

class pcStruct{
    public:
        int x;
        int y;
        char c;
        std::vector<itemStruct> bag; //vector of usable items
        std::vector<pokemon> pokemons; //vector of owned pokemons
        pcStruct(){ //class constructor
            bag.clear();
            pokemons.clear();
        }
};

class npcStruct: public pcStruct{
    public:
        int nextX; //next x pos
        int nextY; //next y pos
        bool isDefeated; //is NPC defeated
        char terrainType;
        pathStruct path; //path
};

//Function to check if spot is occupied by a certain object(building, character, terrain, etc.)
bool spotTaken(char current, char destination[], int size){
    for(int i = 0; i < size; i++){
        if(current == destination[i]){
            return true;
        }
    }
    return false;
}

void initialize_terminal(){
    initscr();
    start_color();
    init_pair(TALL_GRASS, COLOR_GREEN, COLOR_BLACK);
    init_pair(SHORT_GRASS, COLOR_GREEN, COLOR_BLACK);
    init_pair(FOREST, COLOR_GREEN, COLOR_BLACK);
    init_pair(MOUNTAIN, COLOR_YELLOW, COLOR_BLACK);
    init_pair(WATER, COLOR_CYAN, COLOR_BLACK);
    init_pair(PATH, COLOR_MAGENTA, COLOR_BLACK);
    init_pair(POKEMART, COLOR_RED, COLOR_BLACK);
    init_pair(POKECENTER, COLOR_BLUE, COLOR_BLACK);
    init_pair(DEFAULT, COLOR_WHITE, COLOR_BLACK);
    raw();
    noecho();
    curs_set(0);
    keypad(stdscr, TRUE);
}

//Priority Queue for npc movement 
typedef struct pqNode
{
    point data;

    int priority; // lower values = higher priority

    struct pqNode *next;
} pqNode;

// create a new node
pqNode *newNode(point d, int p)
{
    pqNode *temp = (pqNode *)malloc(sizeof(pqNode));
    temp->data = d;
    temp->priority = p;
    temp->next = NULL;

    return temp;
}
// return value at head
point peek(pqNode **head)
{
    return (*head)->data;
}
// removes element with highest priority from list
void pop(pqNode **head)
{
    pqNode *temp = *head;
    (*head) = (*head)->next;
    free(temp);
}
// push according to priority
void push(pqNode **head, point d, int p)
{
    pqNode *start = *head;

    pqNode *temp = newNode(d, p); // create new node

    if ((*head)->priority > p)
    { // new node higher priority than head
        // insert new node before head
        temp->next = *head;
        (*head) = temp;
    }
    else
    {
        while (start->next != NULL && start->next->priority < p)
        {
            start = start->next;
        }

        temp->next = start->next;
        start->next = temp;
    }
}
// check if list empty
int isEmpty(pqNode **head)
{
    return (*head) == NULL;
}


//priority queue for game turn
typedef struct pqGameTurn{
    pcStruct* data;

    int priority; //lower values = higher priority
    int currentTries; //current amount of tries to move to a spot
    struct pqGameTurn* next;
} pqGameTurn;

//Function to create a new node
pqGameTurn* gt_newNode(pcStruct* d, int p){
    pqGameTurn* temp = (pqGameTurn*)malloc(sizeof(pqGameTurn));
    temp->data = d;
    temp->priority = p;
    temp->next = NULL;
    temp->currentTries = 0;
    return temp;
}

pcStruct* gt_peek(pqGameTurn** head){//return value at head
    return(*head)->data;
}

void gt_pop(pqGameTurn** head){//removes element with highest priority from list
    pqGameTurn* temp = *head;
    (*head) = (*head)->next;
    free(temp);
}

void gt_push(pqGameTurn** head, pcStruct* d, int numTries, int p){ //push according to priority
    pqGameTurn* start = *head;

    pqGameTurn* temp = gt_newNode(d, p); //create new node
    temp->currentTries = numTries;

    if((*head)->priority > p){//new node higher priority than head
        //insert new node before head
        temp->next = *head;
        (*head) = temp;
    }
    else{
        while(start->next != NULL && start->next->priority < p){
            start = start->next;
        }

        temp->next = start->next;
        start->next = temp;
    }
}

int gt_isEmpty(pqGameTurn** head){ //check if list empty
    return (*head) == NULL;
}

int gt_getCurrentPriority(pqGameTurn** head){
    return (*head)->priority;
}

int gt_getCurrentTries(pqGameTurn** head){ //return current tries
    return (*head)->currentTries;
}

void gt_setCurrentTries(pqGameTurn** head, int n){ //set current try equals to n
    (*head)->currentTries = n;
}

class worldStruct{
    public:
        char map[ROW_COUNT][COL_COUNT];
        uint8_t height[ROW_COUNT][COL_COUNT];
        int north, south, west, east;
        npcStruct *npcMap[ROW_COUNT][COL_COUNT];
        int num_trainers;
        long hikerDistanceMap[ROW_COUNT][COL_COUNT];
        long rivalDistanceMap[ROW_COUNT][COL_COUNT];
        long swimmerDistanceMap[ROW_COUNT][COL_COUNT];
        pqGameTurn *pqGT;
};

//function to parse csv files
void parseCSV(std::string file){
    std::string path = "";
    char path1[] = "/share/cs327/pokedex";
    char *path2 = getenv("HOME");
    strcat(path2,"/poke327/pokedex");
    char path3[] = "../../../pokedex";

    //Get the path of the database location (1 out of 3 possibilities)
    {
        std::ifstream ifs(path1, std::ifstream::in);
        if(ifs.good()){
            path = path1;
        }
    }

    {
        std::ifstream ifs(path2, std::ifstream::in);
        if(ifs.good()){
            path = path2;
        }
    }

    {
        std::ifstream ifs(path3, std::ifstream::in);
        if(ifs.good()){
            path = path3;
        }
    }

    // cout << path << endl;
    if(path.empty()){ //return with error message if database not found
        cout << "Database not found" << endl;
        return;
    }
    
    path.append("/pokedex/data/csv/");
    path.append(file);
    path.append(".csv");
    std::string line; 
    std::ifstream ifs(path, std::ifstream::in);

    //get number of lines in file
    int lineCount = 0;
    while (std::getline(ifs, line)){
        lineCount++;
    }

    lineCount--; //do not count first row of csv files
    //reset file cursor to start of file
    ifs.clear(); 
    ifs.seekg(0);

    int i = 0;
    bool firstLine = true;
    if(!file.compare("pokemon")){ //case pokemon
        pokemon data[lineCount];
        while (std::getline(ifs, line)){
            if(firstLine){
                firstLine = false;
                continue;
            }

            std::vector<string> values;

            // Create a stringstream from line
            std::stringstream ss(line);
            std::string col;

            // Extract each column data
            while(std::getline(ss, col, ',')){
                if(col.empty()){ //push INT_MAX if empty cell
                    col = "-1";
                }
                values.push_back(col);
            }
            if(ss.eof() && (values.size() == 7)){ //if statement to check if last value is empty
                    col = "-1";
            }

            data[i] = constructPokemon(values);
            i++;
        }

        for(i = 0; i < lineCount; i++){ //add data to vector
            pokemonArray.push_back(data[i]);
        }
    }
    else if(!file.compare("moves")) {//case moves.csv
        moves data[lineCount];
        while (std::getline(ifs, line)){
            if(firstLine){
                firstLine = false;
                continue;
            }

            std::vector<string> values;

            // Create a stringstream from line
            std::stringstream ss(line);
            std::string col;

            // Extract each column data
            while(std::getline(ss, col, ',')){
                if(col.empty()){ //push -1 if empty cell
                    col = "-1";
                }
                values.push_back(col);
            }
            if(ss.eof() && (values.size() == 14)){ //if statement to check if last value is empty
                values.push_back("-1");
            }
            data[i] = constructMoves(values);
            i++;
        }

        for(i = 0; i < lineCount; i++){ //add data to vector
            movesArray.push_back(data[i]);
        }
    }
    else if(!file.compare("pokemon_moves")) {//case pokemon_moves.csv
        std::vector<pokeMoves> data; //use vector for this csv file due to the large size
        while (std::getline(ifs, line)){
            if(firstLine){
                firstLine = false;
                continue;
            }

            std::vector<string> values;

            // Create a stringstream from line
            std::stringstream ss(line);
            std::string col;

            // Extract each column data
            while(std::getline(ss, col, ',')){
                if(col.empty()){ //push -1 if empty cell
                    col = "-1";
                }
                values.push_back(col);
            }
            if(ss.eof() && (values.size() == 5)){ //if statement to check if last value is empty
                values.push_back("-1");
            }
            data.push_back(constructPokeMoves(values));
            i++;
        }

        for(i = 0; i < lineCount; i++){ //copy over data to array
            pokemonMovesArray.push_back(data[i]);
        }
    }
    else if(!file.compare("pokemon_species")) {//case pokemon_species.csv
        pokemon_species data[lineCount];
        while (std::getline(ifs, line)){
            if(firstLine){
                firstLine = false;
                continue;
            }

            std::vector<string> values;

            // Create a stringstream from line
            std::stringstream ss(line);
            std::string col;

            // Extract each column data
            while(std::getline(ss, col, ',')){
                if(col.empty()){ //push -1 if empty cell
                   col = "-1";
                }
                values.push_back(col);
            }
            if(ss.eof() && (values.size() == 19)){ //if statement to check if last value is empty
                values.push_back("-1");
            }
            data[i] = constructSpecies(values);
            i++;
        }

        for(i = 0; i < lineCount; i++){ //copy over data to array
            pokemonSpeciesArray.push_back(data[i]);
        }
    }
    else if(!file.compare("experience")) {//case experience.csv
        experience data[lineCount];
        while (std::getline(ifs, line)){
            if(firstLine){
                firstLine = false;
                continue;
            }

            std::vector<string> values;

            // Create a stringstream from line
            std::stringstream ss(line);
            std::string col;

            // Extract each column data
            while(std::getline(ss, col, ',')){
                if(col.empty()){ //push -1 if empty cell
                    col = "-1";
                }
                values.push_back(col);
            }
            if(ss.eof() && (values.size() == 2)){ //if statement to check if last value is empty
                values.push_back("-1");
            }
            data[i] = constructExperience(values);
            i++;
        }

        for(i = 0; i < lineCount; i++){ //add data to vector
            expArray.push_back(data[i]);
        }
    }
    else if(!file.compare("type_names")) {//case type_names.csv
        type_names data[lineCount];
        while (std::getline(ifs, line)){
            if(firstLine){
                firstLine = false;
                continue;
            }

            std::vector<string> values;

            // Create a stringstream from line
            std::stringstream ss(line);
            std::string col;

            // Extract each column data
            while(std::getline(ss, col, ',')){
                if(col.empty()){ //push -1 if empty cell
                    col = "-1";
                }
                values.push_back(col);
            }
            if(ss.eof() && (values.size() == 2)){ //if statement to check if last value is empty
                values.push_back("-1");
            }
            data[i] = constructTypeNames(values);
            i++;
        }

        for(i = 0; i < lineCount; i++){ //copy over data to array
            typeNamesArray.push_back(data[i]);
        }
    }

    else if(!file.compare("pokemon_stats")) {//case type_names.csv
        pokemon_stats data[lineCount];
        while (std::getline(ifs, line)){
            if(firstLine){
                firstLine = false;
                continue;
            }

            std::vector<string> values;

            // Create a stringstream from line
            std::stringstream ss(line);
            std::string col;

            // Extract each column data
            while(std::getline(ss, col, ',')){
                if(col.empty()){ //push -1 if empty cell
                    col = "-1";
                }
                values.push_back(col);
            }
            if(ss.eof() && (values.size() == 2)){ //if statement to check if last value is empty
                values.push_back("-1");
            }
            data[i] = constructPokeStats(values);
            i++;
        }

        for(i = 0; i < lineCount; i++){ //copy over data to array
            pokemonStatsArray.push_back(data[i]);
        }
    }

    else if(!file.compare("stats")) {//case stats.csv
        stats data[lineCount];
        while (std::getline(ifs, line)){
            if(firstLine){
                firstLine = false;
                continue;
            }

            std::vector<string> values;

            // Create a stringstream from line
            std::stringstream ss(line);
            std::string col;

            // Extract each column data
            while(std::getline(ss, col, ',')){
                if(col.empty()){ //push INT_MAX if empty cell
                    col = "-1";
                }
                values.push_back(col);
            }
            if(ss.eof() && (values.size() == 4)){ //if statement to check if last value is empty
                values.push_back("-1");
            }
            data[i] = constructStat(values);
            i++;
        }

        for(i = 0; i < lineCount; i++){//copy data to array
            statsArray.push_back(data[i]);
        }

        for(i = 0; i < lineCount; i++){//copy data to array
            statsMap.insert(std::pair<std::string, int>(statsArray.at(i).name, statsArray.at(i).id));
        }
    }
    
    else if(!file.compare("pokemon_types")) {//case pokemon_types.csv
        pokemon_types data[lineCount];
        while (std::getline(ifs, line)){
            if(firstLine){
                firstLine = false;
                continue;
            }

            std::vector<string> values;

            // Create a stringstream from line
            std::stringstream ss(line);
            std::string col;

            // Extract each column data
            while(std::getline(ss, col, ',')){
                if(col.empty()){ //push INT_MAX if empty cell
                    col = "-1";
                }
                values.push_back(col);
            }
            if(ss.eof() && (values.size() == 2)){ //if statement to check if last value is empty
                values.push_back("-1");
            }
            data[i] = constructPokeTypes(values);
            i++;
        }

        for(i = 0; i < lineCount; i++){
            pokemonTypesArray.push_back(data[i]);
        }
    }
}

//function to parse specific files in database and store into classes
void getDatabase(){
    std::vector<std::string> files{"pokemon","moves","pokemon_moves","pokemon_species","experience","type_names","pokemon_stats","stats","pokemon_types"};
    
    cout << "Parsing CSV Files to Retrieve Game Data..." << endl;
    int fileSize = files.size();
    for(int i = 0; i < fileSize; i++){
        parseCSV(files[i]);
    }
    cout << "Game Data Retrieved Successfully!" << endl;
    usleep(1250000);
}

util5_database.h
/*
Header file to store csv data structs
*/
#include <string>
#include <vector>
#include <iostream>
#include <cstring>
#include <cstdlib>
#include <climits>
#include <map>

#ifndef UTIL5_DATABASE_H
# define UTIL5_DATABASE_H
# endif

using namespace std;

#define NUM_STATS 8

//******************** MOVES.CSV ********************//
class moves{
    public:
        int id;
        std::string name;
        int generation_id;
        int type_id;
        int power;
        int pp;
        int accuracy;
        int priority;
        int target_id;
        int damage_class_id;
        int effect_id;
        int effect_chance;
        int contest_type_id;
        int contest_effect_id;
        int super_contest_effect_id;
        friend std::ostream &operator<<(std::ostream& o, const moves &m);
};

moves constructMoves(std::vector<std::string> values){ //moves constructor
    moves move;

    move.id = std::stoi(values[0]);
    move.name = values[1];
    move.generation_id = std::stoi(values[2]);
    move.type_id = std::stoi(values[3]);
    move.power = std::stoi(values[4]);
    move.pp = std::stoi(values[5]);
    move.accuracy = std::stoi(values[6]);
    move.priority = std::stoi(values[7]);
    move.target_id = std::stoi(values[8]);
    move.damage_class_id = std::stoi(values[9]);
    move.effect_id = std::stoi(values[10]);
    move.effect_chance = std::stoi(values[11]);
    move.contest_type_id = std::stoi(values[12]);
    move.contest_effect_id = std::stoi(values[13]);
    move.super_contest_effect_id = std::stoi(values[14]);

    return move;
}

std::ostream &operator<<(std::ostream &o, const moves &m){ //print function for moves.csv
    o << ((m.id != -1) ? std::to_string(m.id) : "") << ",";
    o << ((m.name != "-1") ? m.name : "") << ",";
    o << ((m.generation_id != -1) ? std::to_string(m.generation_id) : "") << ",";
    o << ((m.type_id != -1) ? std::to_string(m.type_id) : "") << ",";
    o << ((m.power != -1) ? std::to_string(m.power) : "") << ",";
    o << ((m.pp != -1) ? std::to_string(m.pp) : "") << ",";
    o << ((m.accuracy != -1) ? std::to_string(m.accuracy) : "") << ",";
    o << ((m.priority != -1) ? std::to_string(m.priority) : "") << ",";
    o << ((m.target_id != -1) ? std::to_string(m.target_id) : "") << ",";
    o << ((m.damage_class_id != -1) ? std::to_string(m.damage_class_id) : "") << ",";
    o << ((m.effect_id != -1) ? std::to_string(m.effect_id) : "") << ",";
    o << ((m.effect_chance != -1) ? std::to_string(m.effect_chance) : "") << ",";
    o << ((m.contest_type_id != -1) ? std::to_string(m.contest_type_id) : "") << ",";
    o << ((m.contest_effect_id != -1) ? std::to_string(m.contest_effect_id) : "") << ",";
    o << ((m.super_contest_effect_id != -1) ? std::to_string(m.super_contest_effect_id) : "");
    return o;
}

//******************** POKEMON_MOVES.CSV ********************//
class pokeMoves{
    public:
        int id;
        int version_group_id;
        int move_id;
        int pokemon_move_method_id;
        int level;
        int order;
        friend std::ostream &operator<<(std::ostream& o, const pokeMoves &pm);
};

pokeMoves constructPokeMoves(std::vector<std::string> values){ //pokemon_moves constructor
    pokeMoves pokeMove;

    pokeMove.id = std::stoi(values[0]);
    pokeMove.version_group_id = std::stoi(values[1]);
    pokeMove.move_id = std::stoi(values[2]);
    pokeMove.pokemon_move_method_id = std::stoi(values[3]);
    pokeMove.level = std::stoi(values[4]);
    pokeMove.order = std::stoi(values[5]);

    return pokeMove;
}

std::ostream &operator<<(std::ostream &o, const pokeMoves &pm){ //print function for pokemon_moves.csv
    o << ((pm.id != -1) ? std::to_string(pm.id) : "") << ",";
    o << ((pm.version_group_id != -1) ? std::to_string(pm.version_group_id) : "") << ",";
    o << ((pm.move_id != -1) ? std::to_string(pm.move_id) : "") << ",";
    o << ((pm.pokemon_move_method_id != -1) ? std::to_string(pm.pokemon_move_method_id) : "") << ",";
    o << ((pm.level != -1) ? std::to_string(pm.level) : "") << ",";
    o << ((pm.order != -1) ? std::to_string(pm.order) : "");
    
    return o;
}

//******************** POKEMON_SPECIES.CSV ********************//
class pokemon_species{
    public:
        int id;
        std::string name;
        int generation_id;
        int evolves_from_species_id;
        int evolution_chain_id;
        int color_id;
        int shape_id;
        int habitat_id;
        int gender_rate;
        int capture_rate;
        int base_happiness;
        int is_baby;
        int hatch_counter;
        int has_gender_differences;
        int growth_rate_id;
        int forms_switchable;
        int is_legendary;
        int is_mythical;
        int order;
        int conquest_order;
        friend std::ostream &operator<<(std::ostream& o, const pokemon_species &ps);
};

pokemon_species constructSpecies(std::vector<std::string> values){ //pokemon_species constructor
    pokemon_species species;

    species.id = stoi(values[0]);
    species.name = values[1];
    species.generation_id = stoi(values[2]);
    species.evolves_from_species_id = stoi(values[3]);
    species.evolution_chain_id = stoi(values[4]);
    species.color_id = stoi(values[5]);
    species.shape_id = stoi(values[6]);
    species.habitat_id = stoi(values[7]);
    species.gender_rate = stoi(values[8]);
    species.capture_rate = stoi(values[9]);
    species.base_happiness = stoi(values[10]);
    species.is_baby = stoi(values[11]);
    species.hatch_counter = stoi(values[12]);
    species.has_gender_differences = stoi(values[13]);
    species.growth_rate_id = stoi(values[14]);
    species.forms_switchable = stoi(values[15]);
    species.is_legendary = stoi(values[16]);
    species.is_mythical = stoi(values[17]);
    species.order = stoi(values[18]);
    species.conquest_order = stoi(values[19]);

    return species;
}

std::ostream &operator<<(std::ostream &o, const pokemon_species &ps){ //print function for pokemon_species.csv
    o << ((ps.id != -1) ? std::to_string(ps.id) : "") << ",";
    o << ((ps.name != "-1") ? ps.name : "") << ",";
    o << ((ps.generation_id != -1) ? std::to_string(ps.generation_id) : "") << ",";
    o << ((ps.evolves_from_species_id != -1) ? std::to_string(ps.evolves_from_species_id) : "") << ",";
    o << ((ps.evolution_chain_id != -1) ? std::to_string(ps.evolution_chain_id) : "") << ",";
    o << ((ps.color_id != -1) ? std::to_string(ps.color_id) : "") << ",";
    o << ((ps.shape_id != -1) ? std::to_string(ps.shape_id) : "") << ",";
    o << ((ps.habitat_id != -1) ? std::to_string(ps.habitat_id) : "") << ",";
    o << ((ps.gender_rate != -1) ? std::to_string(ps.gender_rate) : "") << ",";
    o << ((ps.capture_rate != -1) ? std::to_string(ps.capture_rate) : "") << ",";
    o << ((ps.base_happiness != -1) ? std::to_string(ps.base_happiness) : "") << ",";
    o << ((ps.is_baby != -1) ? std::to_string(ps.is_baby) : "") << ",";
    o << ((ps.hatch_counter != -1) ? std::to_string(ps.hatch_counter) : "") << ",";
    o << ((ps.has_gender_differences != -1) ? std::to_string(ps.has_gender_differences) : "") << ",";
    o << ((ps.growth_rate_id != -1) ? std::to_string(ps.growth_rate_id) : "") << ",";
    o << ((ps.forms_switchable != -1) ? std::to_string(ps.forms_switchable) : "") << ",";
    o << ((ps.is_legendary != -1) ? std::to_string(ps.is_legendary) : "") << ",";
    o << ((ps.is_mythical != -1) ? std::to_string(ps.is_mythical) : "") << ",";
    o << ((ps.order != -1) ? std::to_string(ps.order) : "") << ",";
    o << ((ps.conquest_order != -1) ? std::to_string(ps.conquest_order) : "");
    return o;
}

//******************** EXPERIENCE.CSV ********************//
class experience{
    public:
        int growth_rate_id;
        int level;
        int exp;
        friend std::ostream &operator<<(std::ostream& o, const experience &e);
};

experience constructExperience(std::vector<std::string> values){ //experience constructor
    experience exp;

    exp.growth_rate_id = stoi(values[0]);
    exp.level = stoi(values[1]);
    exp.exp = stoi(values[2]);

    return exp;
}

std::ostream &operator<<(std::ostream &o, const experience &e){ //print function for exp.csv
    o << ((e.growth_rate_id != -1) ? std::to_string(e.growth_rate_id) : "") << ",";
    o << ((e.level != -1) ? std::to_string(e.level) : "") << ",";
    o << ((e.exp != -1) ? std::to_string(e.exp) : "");
    return o;
}

//******************** TYPE_NAMES.CSV ********************//
class type_names{
    public:
        int type_id;
        int local_language_id;
        std::string name;
        friend std::ostream &operator<<(std::ostream& o, const type_names &tn);
};

type_names constructTypeNames(std::vector<std::string> values){ //type name constructor
    type_names typeName;

    typeName.type_id = stoi(values[0]);
    typeName.local_language_id = stoi(values[1]);
    typeName.name = values[2];

    return typeName;
}

std::ostream &operator<<(std::ostream &o, const type_names &tn){ //print function for type_name.csv
    o << ((tn.type_id != -1) ? std::to_string(tn.type_id) : "") << ",";
    o << ((tn.local_language_id != -1) ? std::to_string(tn.local_language_id) : "") << ",";
    o << ((tn.name != "-1") ? tn.name : "");
    return o;
}

//******************** POKEMON_STATS.CSV ********************//
class pokemon_stats{
    public:
        int pokemon_id;
        int stat_id;
        int base_stat;
        int effort;
        friend std::ostream &operator<<(std::ostream& o, const pokemon_stats &pks);
};

pokemon_stats constructPokeStats(std::vector<std::string> values){ //pokemon stat constructor
    pokemon_stats pokeStat ;

    pokeStat.pokemon_id = stoi(values[0]);
    pokeStat.stat_id = stoi(values[1]);
    pokeStat.base_stat = stoi(values[2]);
    pokeStat.effort = stoi(values[3]);

    return pokeStat;
}

std::ostream &operator<<(std::ostream &o, const pokemon_stats &pks){ //print function for pokemon_stats.csv
    o << ((pks.pokemon_id != -1) ? std::to_string(pks.pokemon_id) : "") << ",";
    o << ((pks.stat_id != -1) ? std::to_string(pks.stat_id) : "") << ",";
    o << ((pks.base_stat != -1) ? std::to_string(pks.base_stat) : "") << ",";
    o << ((pks.effort != -1) ? std::to_string(pks.effort) : "") << ",";

    return o;
}

//******************** STATS.CSV ********************//
class stats{
    public:
        int id;
        int damage_class_id;
        std::string name;
        int is_battle_only;
        int game_index;
        friend std::ostream &operator<<(std::ostream& o, const stats &s);
};

stats constructStat(std::vector<std::string> values){ //stats constructor
    stats stat;

    stat.id = stoi(values[0]);
    stat.damage_class_id = stoi(values[1]);
    stat.name = values[2];
    stat.is_battle_only = stoi(values[3]);
    stat.game_index = stoi(values[4]);

    return stat;
}

std::ostream &operator<<(std::ostream &o, const stats &s){ //print function for stats.csv
    o << ((s.id != -1) ? std::to_string(s.id) : "") << ",";
    o << ((s.damage_class_id != -1) ? std::to_string(s.damage_class_id) : "") << ",";
    o << ((s.name != "-1") ? s.name : "") << ",";
    o << ((s.is_battle_only != -1) ? std::to_string(s.is_battle_only) : "") << ",";
    o << ((s.game_index != -1) ? std::to_string(s.game_index) : "") << ",";
    return o;
}

//******************** POKEMON_TYPE.CSV ********************//
class pokemon_types{
    public:
        int pokemon_id;
        int type_id;
        int slot;
        friend std::ostream &operator<<(std::ostream& o, const stats &p);
};

pokemon_types constructPokeTypes(std::vector<std::string> values){ //pokemon type constructor
    pokemon_types pokeType ;

    pokeType.pokemon_id = stoi(values[0]);
    pokeType.type_id = stoi(values[1]);
    pokeType.slot = stoi(values[2]);

    return pokeType;
}

std::ostream &operator<<(std::ostream &o, const pokemon_types &p){ //print function for pokemon_types.csv
    o << ((p.pokemon_id != -1) ? std::to_string(p.pokemon_id) : "") << ",";
    o << ((p.type_id != -1) ? std::to_string(p.type_id) : "") << ",";
    o << ((p.slot != -1) ? std::to_string(p.slot) : "") << ",";

    return o;
}
//******************** POKEMON.CSV ********************//
class pokemon{
    public:
        int id;
        std::string name;
        int species;
        int height;
        int weight;
        int base_experience;
        int order;
        int is_default;
        int level;
        int gender; // 0 - male, 1 - female
        int is_shiny; //0 - no, 1 - yes
        std::vector<moves> move_set;
        int currHP;
        int base_stats[NUM_STATS]; //HP, Attack, Defense, Special Attack, Special Defense, Speed, Accuracy, Evasion
        int stats[NUM_STATS]; //HP, Attack, Defense, Special Attack, Special Defense, Speed, Accuracy, Evasion (Current stats)
        int stats_iv[NUM_STATS]; //HP, Attack, Defense, Special Attack, Special Defense, Speed, Accuracy, Evasion (Individual Value)
        friend std::ostream &operator<<(std::ostream& o, const pokemon &p);
};

pokemon constructPokemon(std::vector<std::string> values){ //pokemon constructor
    pokemon poke;

    poke.id = stoi(values[0]);
    poke.name = values[1];
    poke.species = stoi(values[2]);
    poke.height = stoi(values[3]);
    poke.weight = stoi(values[4]);
    poke.base_experience = stoi(values[5]);
    poke.order = stoi(values[6]);
    poke.is_default = stoi(values[7]);

    return poke;
}

std::ostream &operator<<(std::ostream &o, const pokemon &p){ //print function for pokemon.csv
    o << ((p.id != -1) ? std::to_string(p.id) : "") << ",";
    o << ((p.name != "-1") ? p.name : "") << ",";
    o << ((p.species != -1) ? std::to_string(p.species) : "") << ",";
    o << ((p.height != -1) ? std::to_string(p.height) : "") << ",";
    o << ((p.weight != -1) ? std::to_string(p.weight) : "") << ",";
    o << ((p.base_experience != -1) ? std::to_string(p.base_experience) : "") << ",";
    o << ((p.order != -1) ? std::to_string(p.order) : "") << ",";
    o << ((p.is_default != -1) ? std::to_string(p.is_default) : "");
    return o;
}
